### DISCUSSION TITLE
Maui.Graphics vs SkiaSharp

### DISCUSSION THREAD

**Discussion by bmitc** [dotnet/System.Graphics#272] [General]
Hello! I have been sketching out a cross-platform graphics oriented project, intending to use SkiaSharp for the graphics and text capability. But I came across this today. So, I have some questions. :)

1. What does Maui.Graphics offer over SkiaSharp? Or, why use Maui.Graphics over SkiaSharp?
2. Possibly depending on the answer to (1), why not standardize and concentrate effort on SkiaSharp?

Thanks!

**Comment by xtuzy** [comment]
the size of skiasharp is big, maui want little size, they want more "native". skia is "full" graphics library, maui.graphic not .

**Reply by bmitc** [reply]
I'm not sure I buy that. [SkiaSharp is a supported backend](https://github.com/dotnet/Microsoft.Maui.Graphics#supported-platforms) for Maui.Graphics.

**Reply by xtuzy** [reply]
🤣i don't know how to choose, android canvas api is meet your demand? i have a plan to base on android canvas api wrap skiasharp and win2d api, only the user of android more sensitive to app size. 
i like skiasharp run everywhere, but winui3 don't have sk gpu view

**Reply by bmitc** [reply]
I only care about cross-platform desktop and not mobile or web. 🙂 Right now SkiaSharp plus GLFW is the only choice if one wants to avoid Qt, GTK, and SDL and also MAUI, Uno, and Avalonia. Maui.Graphics could be a choice, but that's why I am concerned with it competing with SkiaSharp. I wish there was just a single library that has solid attention and usage.

**Reply by xtuzy** [reply]
See https://github.com/dotnet/maui/issues/7901, Maui.Graphics have better support, but i don't know 
"Remove non-Maui platforms (SharpDX, UWP, JS)" mean what at https://github.com/dotnet/maui/pull/8147

**Comment by bmitc** [comment]
Does anyone official from .NET MAUI have some thoughts on this? There's already so much fracturing going on for .NET GUIs and graphics. There's this library which is apparently experimental, SkiaSharp which needs maintenance help, and even [Avalonia is now developing their own Skia bindings because of that](https://github.com/AvaloniaUI/Avalonia/discussions/7629).

The *why* for this library is confusing, especially since both Maui.Graphics and SkiaSharp are ostensibly owned by Microsoft. (Then of course there's Win2D on Windows.)

**Reply by jonlipsky** [reply]
@bmitc As the original author of Maui.Graphics, I can try explain why I feel it needs to exist, and why SkiaSharp is not the end-all-be-all solution.

First, and foremost, the my goal with Maui.Graphics is to allow access to the *native* graphics platform with a consistent API.  When I am writing an iOS or Mac app, I want my graphics to be rendered with CoreGraphics.  When I write a Windows app, I want my graphics to be rendered with Direct 2D.  I want my font rendering to be consistent to the user, I want anti-aliasing to be consistent to the user, etc...

Second, because of the types of apps that I typically write, I may want access use multiple graphics engines on a single platform.  An example of this, is that Windows does not provide a native API to create PDF files; however Skia (and SkiaSharp) does.  For "on-screen rendering" I may want to use Direct 2D; however, to create a PDF, I may want to use SkiaSharp.  With that said, I've written (closed source, for now) Maui.Graphics wrappers around another PDF engine (Xfinium.Pdf) when I needed access to advanced PDF features that SkiaSharp doesn't/can't provide.  Did I have to change my Maui.Graphics code that generated the PDF?  No.  I just had to change the backend implementation.

The choice of the best solution depends on the the type of app you are writing, and the goals/priorities of the company owning the app.  With Maui.Graphics, I can choose the rendering engine that best fits my needs.

**Comment by jevonsflash** [comment]
Maui.Graphics is certainly lighter, but there are two reasons to use SkiaSharp rendering:

First, SkiaSharp supports more platforms and is more consistent across platforms

Second, when upgrading from the xamarin.forms project to the Maui platform, if using an very-old xmaarin.forms library (perhaps around 2.X ?), it seems like the SkiaSharp is the only choice..
For others reason, Maui.Graphics is fine
